src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import algebra.group.units
+
+/-!
+# Groups with an adjoined zero element
+
+This file describes structures that are not usually studied on their own right in mathematics,
+namely a special sort of monoid: apart from a distinguished “zero element” they form a group,
+or in other words, they are groups with an adjoined zero element.
+
+Examples are:
+
+* division rings;
+* the value monoid of a multiplicative valuation;
+* in particular, the non-negative real numbers.
+
+## Main definitions
+
+* `group_with_zero`
+* `comm_group_with_zero`
+
+## Implementation details
+
+As is usual in mathlib, we extend the inverse function to the zero element,
+and require `0⁻¹ = 0`.
+
+-/
+
+set_option old_structure_cmd true
+
+section prio
+set_option default_priority 10 -- see Note [default priority]
+
+/-- A type `M` is a “monoid with zero” if it is a monoid with zero element. -/
+class monoid_with_zero (G : Type*) extends monoid G, mul_zero_class G.
+
+/-- A type `M` is a commutative “monoid with zero”
+if it is a commutative monoid with zero element. -/
+class comm_monoid_with_zero (G : Type*) extends monoid G, mul_zero_class G.
+
+/-- A type `G` is a “group with zero” if it is a monoid with zero element (distinct from `1`)
+such that every nonzero element is invertible.
+The type is required to come with an “inverse” function, and the inverse of `0` must be `0`.
+
+Examples include division rings and the ordered monoids that are the
+target of valuations in general valuation theory.-/
+class group_with_zero (G : Type*)
+  extends monoid_with_zero G, has_inv G, zero_ne_one_class G :=
+(inv_zero : (0 : G)⁻¹ = 0)
+(mul_inv_cancel : ∀ a:G, a ≠ 0 → a * a⁻¹ = 1)
+
+/-- A type `G` is a commutative “group with zero”
+if it is a commutative monoid with zero element (distinct from `1`)
+such that every nonzero element is invertible.
+The type is required to come with an “inverse” function, and the inverse of `0` must be `0`. -/
+class comm_group_with_zero (G : Type*) extends comm_monoid G, group_with_zero G.
+
+/-- The division operation on a group with zero element. -/
+instance group_with_zero.has_div {G : Type*} [group_with_zero G] :
+  has_div G := ⟨λ g h, g * h⁻¹⟩
+
+end prio
+
+lemma div_eq_mul_inv {G : Type*} [group_with_zero G] {a b : G} :
+  a / b = a * b⁻¹ := rfl
+
+namespace gwz
```

Comment:
You lose dot notation with this character-saving namespace, right?

Reply:
Yes, but you don't want to use dot notation anyway, because it's an instance, so you don't see the variable that you would want to put the dot-notation on.
At the same time, you probably don't want to open this namespace, because you'll get many conflicting names after that. Hence I chose for a short prefix as namespace.

Reply:
Still, needlessly obfuscatory.

Reply:
I've removed the `gwz` namespace.